include/misc/rb.h

```diff
@@ -97,15 +97,55 @@ int rb_contains(struct rbtree *tree, struct rbnode *node);
 /**
  * @brief Walk/enumerate a rbtree
  *
- * Very simple recursive enumeration implementation.  A rather more
- * subtle (have to alloca() a stack to manage manually) iterative one
- * is possible that uses a FOREACH-style macro API, but this is good
- * enough for many purposes and much smaller.
+ * Very simple recursive enumeration.  Low code size, but requiring a
+ * separate function can be clumsy for the user and there is no way to
+ * break out of the loop early.  See RB_FOR_EACH for an iterative
+ * implementation.
  */
 static inline void rb_walk(struct rbtree *tree, rb_visit_t visit_fn,
 			   void *cookie)
 {
 	_rb_walk(tree->root, visit_fn, cookie);
 }
 
+struct _rb_foreach {
+	struct rbnode **stack;
+	char *is_left;
+	int top;
+};
+
+#define _RB_FOREACH_INIT(tree, node) {					\
+	.stack   = alloca((tree)->max_depth * sizeof(struct rbnode *)), \
+	.is_left = alloca((tree)->max_depth * sizeof(char)), 		\
```

Comment:
is_left could be a bitmask to reduce stack usage

Reply:
Thought about that, but it would add significant code size (I dunno, maybe 20-30 bytes?) to do the twiddling just to save 17.5% of the allocated stack space.    So that becomes worth it only when the tree depth grows beyond ~20ish, which corresponds to a tree with about a thousand entries.  Only worth it in the extremes of scalability where frankly we probably aren't sweating a few bytes of stack anyway.